### ISSUE TITLE
Higher level API options

### ISSUE THREAD

**Issue by trusktr** [BearishSun/BansheeEngine#161]
I'm looking into open-source 3D libs like Ogre and Banshee. Just curious: is there a c++ API that's as high level as Oger's in Banshee (3rd party stuff is okay)? 

I'm looking for something as high level as Three.js, but in "native" land.

**Comment by BearishSun** [comment]
I'm not familiar with Three.js, but Banshee does provide both a high level and low level C++ API.

**Comment by w-flo** [comment]
See the examples, e.g.:
https://github.com/BearishSun/BansheeEngine/blob/master/Source/Examples/ExampleGettingStarted/Main.cpp

Also, the manuals:
http://docs.banshee3d.com/Native/manuals.html

I'd say BearishSun did an excellent job, but I'm not sure about the high-levelness of Three.js, so maybe you're looking for something different. Just note that Banshee sees lots of changes and bug fixes, and there's no stable release yet, so you can't really rely on everything to work perfectly.

**Comment by trusktr** [comment]
I'm looking to see if I can find something as simple as the Three.js API, but in C/C++.

For example, see these demos and click the View Source button and look at the JS:

- https://threejs.org/examples/#webgl_geometry_cube
- https://threejs.org/examples/#webgl_loader_collada_skinning
- https://threejs.org/examples/#webgl_lights_hemisphere
- https://threejs.org/examples/#webgl_geometry_minecraft_ao

The API could be even higher level and easier which is why I'm working on infamous.io.

So far the simplest top-level API I've found is in Ogre3D. F.e., look at these two tutorials:

- https://ogrecave.github.io/ogre/api/1.10/tut_FirstScene.html
- https://ogrecave.github.io/ogre/api/1.10/tut_LightsCamerasShadows.html

Three.js and Ogre are focused on letting us describe a scene in a minimal way. Do you have any examples like these ones for Banshee?

**Comment by trusktr** [comment]
Okay, I just looked in the manual, and some things seem easy to do. Maybe it would be great to have some complete whole simple examples like Ogre and Three.js have.

**Comment by trusktr** [comment]
Three.js also recently added this [incredibly awesome API](https://github.com/mrdoob/three.js/pull/13070) for composing shaders out of various nodes (similar to audio programs that generate different sound based on a configuration of connected Nodes).

Think of it as the programmatic way to do this with Shaders:

![](https://cloud.githubusercontent.com/assets/502810/11191568/0275e236-8c83-11e5-8b18-060397b6cb73.jpg)

![](https://cloud.githubusercontent.com/assets/502810/11191646/73081230-8c83-11e5-92b6-ed6921145abe.jpg)

Making shader effects highly compostable. Two years in the making ([see here for more info](https://github.com/mrdoob/three.js/issues/7522)).

**Comment by trusktr** [comment]
FWIW, I clicked on "Docs" on the website, and the User Manuals section said "coming soon" so I didn't see those examples before

**Comment by w-flo** [comment]
Yeah, those manuals I linked are listed as "Core Manuals". High-Level manuals (referring to editor / C# scripting) are "coming soon". Just like you, I'm more interested in the C++ API.

That shader composition feature looks interesting. As far as I know (and I have only recently discovered Banshee 3D), RenderBest (renderer plugin) uses some kind of tree structure to compose different rendering steps, too, but it's probably not as generic and there's no API to influence it. See https://github.com/BearishSun/BansheeEngine/blob/master/Source/RenderBeast/BsRenderCompositor.h - mabye at some point in the future that could be made more generic and allow API users to plug-in their own rendering steps or re-arrange existing steps. Though I'm not sure how useful it actually is in practice, when all you want to do is simply enable or disable a certain feature or post-processing step. :-)
